### ISSUE TITLE
Empty cells in functional abundance table

### ISSUE THREAD

**Issue by EricRaes** [jtamames/SqueezeMeta#122]
Hi guys,

Hope you are having a good week.

I was hoping you could shed some light on my question. I did an sqm_reads run, followed by a sqmreads2tables.pys and loaded this output into SQMtools.

I export the functional abundance table via: 
B1_fun = B1$functions$KEGG_paths$abund
rownames(B1_fun) = B1_prok$misc$KEGG_paths[rownames(B1_fun)]
exportTable(B1_fun, 'B1_prok_KEGG_paths.tsv')

When I look at the file there are a number of empty cells; e.g.,

KEGG | Sample_A1
          | 59
          | 9958
          | 27994
          | 9
          | 8
          | 14
          | 7
          | 189
          | 1149
          | 127
          | 5
          | 5

Would you have an idea why there are empty cells? And maybe why they aren't aggregated in the "NA" grouping?

Many thanks in advance!
Cheers,
Eric

**Comment by fpusan** [comment]
Yeah, that looks like it should not be happening.
Could you compress the output directory from `sqmreads2tables.py` and attach it here?

**Comment by EricRaes** [comment]
Hey Fernando,

Please find a zipped output attached.

[Derwent_Sites_Tables_SqueezeMeta_U2.zip](https://github.com/jtamames/SqueezeMeta/files/4817661/Derwent_Sites_Tables_SqueezeMeta_U2.zip)

Also, a minor point but the '....prokfilter.abund.tsv' file contains non-proks, the tax_mode = "prokfilter" in the SQMtools package then doesn't work.

Thanks a lot for your help!

**Comment by fpusan** [comment]
The prokfilter.abund file is not about having only prokariotes, but about applying identity-based taxonomic filters only to prokaryotic taxa.

See #79 for further discussion (the --euk flag would make the same effect as using the prokfilter files).

I'll take a look to your files!

**Comment by fpusan** [comment]
I can not reproduce the issue with the code you sent me. You seem to have two different objects `B1` and `B1_prok`. Also, `B1$functions$KEGG_paths` should not exist unless you created manually at some point (it would normally be `B1$functions$KEGG`).

I still think there might be a bug on our part, but first I'd like to be sure of what you wanted to do exactly.

**Comment by EricRaes** [comment]
Hey Fernando,

Thanks again for your swift reply and help!

I loaded the tabular files from the sqmreads2tables.py output into R using SQMtools. I then wanted to export a functional table and explore it further, as a sanity check for myself.

Apologies for the previous mistake in the R code, yet when I use the below I still get empty cells:

[results_B1.zip](https://github.com/jtamames/SqueezeMeta/files/4819015/results_B1.zip)

library(SQMtools)

B1 = loadSQMlite('~/results_B1/tables')

ta = B1$functions$KEGG$abund
rownames(ta) = B1$misc$KEGG_names[rownames(ta)]
exportTable(ta, 'B1_test_to_see_if_there_are_empties.tsv')
write.csv(ta, "B1_test_to_see_if_there_are_empties.csv")

When you open the .csv file and do a quick filter in excel you see about 10 empty (Blank) cells.

Many thanks for your help!

**Comment by fpusan** [comment]
Ok, I think I might know what's going on.

Some function names are missing (actually due to more than one reason). I'll work on a patch, but for now you can just do this.

```
library(SQMtools)

B1 = loadSQMlite('~/results_B1/tables')

ta = B1$functions$KEGG$abund

### temporarily fix missing KEGG names
all_names = B1$misc$KEGG_names
all_names[nchar(all_names )==0] = names(all_names )[nchar(all_names )==0] # Multi KEGG (all) or missing in data/keggfun2.txt
missing_names = setdiff(rownames(ta), names(B1$misc$KEGG_names))          # Members of multi KEGG which do not appear alone (and "Unclassified")
names(missing_names) = missing_names 
all_names = c(all_names, missing_names)

#Go on
rownames(ta) = all_names[rownames(ta)]
exportTable(ta, 'B1_somehow_better.tsv')
```

This is not a perfect solution (when a name is missing, you will get the KEGG code instead) but at least returns no empty row names or NAs.

I'll let you know when I have a better solution.

**Comment by EricRaes** [comment]
Thanks Fernando!

**Comment by fpusan** [comment]
Hopefully fixed in  c222701
